File: Algorithms/src/1701. Average Waiting Time/Code/solution.java
class Solution {
    public double averageWaitingTime(int[][] customers) {
        // Variable to store the total waiting time for all customers
        long totalWaitingTime = 0;

        // Variable to track the current time in the restaurant
        int currentTime = 0;

        // Iterate over each customer in the list of customers
        for (int[] customer : customers) {
            // Get the arrival time and cooking time for the current customer
            int arrivalTime = customer[0];
            int cookingTime = customer[1];

            // Update the current time to be the later of the current time or the customer's
            // arrival time,
            // then add the cooking time. This represents the time at which the current
            // customer's order is finished.
            currentTime = Math.max(currentTime, arrivalTime) + cookingTime;

            // Calculate the waiting time for the current customer by subtracting their
            // arrival time
            // from the time their order is finished, and add it to the total waiting time.
            totalWaitingTime += currentTime - arrivalTime;
        }

        // Calculate and return the average waiting time by dividing the total waiting
        // time
        // by the number of customers. Cast totalWaitingTime to double to ensure the
        // result is a double.
        return (double) totalWaitingTime / customers.length;
    }
}

File: Algorithms/src/1701. Average Waiting Time/Code/solution.cpp
class Solution
{
public:
    double averageWaitingTime(vector<vector<int>> &customers)
    {
        // Initialize a variable to store the total waiting time of all customers
        long long total_waiting_time = 0;

        // Initialize a variable to keep track of the current time in the restaurant
        int current_time = 0;

        // Iterate through each customer in the customers list
        for (const auto &customer : customers)
        {
            // Extract the arrival time and cooking time for the current customer
            int arrival_time = customer[0];
            int cooking_time = customer[1];

            // Update the current time to either the customer's arrival time or
            // the end of the previous customer's cooking time, whichever is later,
            // and add the current customer's cooking time
            current_time = max(current_time, arrival_time) + cooking_time;

            // Calculate the waiting time for the current customer by subtracting
            // the arrival time from the current time (which now includes their cooking time)
            total_waiting_time += current_time - arrival_time;
        }

        // Calculate and return the average waiting time by dividing the total
        // waiting time by the number of customers
        return (double)total_waiting_time / customers.size();
    }
};
